servicetalk-opentracing-zipkin-publisher/src/main/java/io/servicetalk/opentracing/zipkin/publisher/reporter/UdpReporter.java

```diff
@@ -0,0 +1,229 @@
+            this.loggerName = loggerName;
+            return this;
+        }
+
+        /**
+         * Builds a new {@link UdpReporter} instance with this builder's options.
+         *
+         * @return a new {@link UdpReporter}
+         */
+        public UdpReporter build() {
+            return new UdpReporter(this);
+        }
+    }
+
+    private Bootstrap buildBootstrap(EventLoopGroup group, Codec codec, SocketAddress collectorAddress,
+                                     @Nullable String loggerName) {
+        if (!(collectorAddress instanceof InetSocketAddress)) {
+            throw new IllegalArgumentException("collectorAddress " + collectorAddress +
+                    " is invalid for UDP");
+        }
+        return new Bootstrap()
+                .group(group)
+                .channel(datagramChannel(group))
+                .option(RCVBUF_ALLOCATOR, DEFAULT_RECV_BUF_ALLOCATOR)
+                .handler(new ChannelInitializer<Channel>() {
+                    @Override
+                    protected void initChannel(final Channel ch) {
+                        if (loggerName != null) {
+                            ch.pipeline().addLast(new LoggingHandler(loggerName));
+                        }
+                        ch.pipeline().addLast(new ChannelOutboundHandlerAdapter() {
+                            @Override
+                            public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) {
+                                if (msg instanceof Span) {
+                                    byte[] bytes = codec.encoder.encode((Span) msg);
+                                    ByteBuf buf = ctx.alloc().buffer(bytes.length).writeBytes(bytes);
+                                    ctx.write(new DatagramPacket(buf, (InetSocketAddress) collectorAddress), promise);
+                                } else {
+                                    ctx.write(msg, promise);
+                                }
+                            }
+                        });
+                    }
+                });
+    }
+
+    /**
+     * Non-blocking report method.
+     *
+     * @param span the span to report
+     */
+    @Override
+    public void report(final Span span) {
+        if (closed) {
+            throw new IllegalStateException("closed");
+        }
+        channel.writeAndFlush(span);
+    }
+
+    /**
+     * Blocking close method delegates to {@link #closeAsync()}).
+     */
+    @Override
+    public void close() {
+        try {
+            closeAsync().toFuture().get();
+        } catch (InterruptedException | ExecutionException e) {
+            e.printStackTrace();
+        }
+    }
+
+    /**
+     * Marks the {@link UdpReporter} as closed to prevent accepting any more {@link Span}s.
+     *
+     * @return a {@link Completable} that is completed when close is done.
+     */
+    @Override
+    public Completable closeAsync() {
+        closed = true;
```

Comment:
We still need to close the `channel`. You can make [this class](https://github.com/apple/servicetalk/blob/master/servicetalk-transport-netty-internal/src/main/java/io/servicetalk/transport/netty/internal/NettyFutureCompletable.java#L28) `public` and use it here:

```java
return new NettyFutureCompletable(channel::closeFuture);
```

Better to create an keep this `NettyFutureCompletable` instance in the constructor.